File: src/main/java/org/taktik/icure/dao/PatientDAO.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.dao;

import org.ektorp.support.View;
import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.db.PaginationOffset;
import org.taktik.icure.entities.Patient;

import java.util.Collection;
import java.util.List;

public interface PatientDAO extends GenericDAO<Patient> {

	List<String> listIdsByHcPartyAndName(String searchString, String healthcarePartyId);
	List<String> listIdsOfHcPartyAndName(String searchString, String healthcarePartyId);
	List<String> listIdsByHcPartyAndSsin(String ssin, String healthcarePartyId);
	List<String> listIdsOfHcPartyAndSsin(String ssin, String healthcarePartyId);
	List<String> listIdsByActive(boolean active, String healthcarePartyId);

	List<Patient> listOfMergesAfter(Long date);

	Integer countByHcParty(String healthcarePartyId);

	Integer countOfHcParty(String healthcarePartyId);

	List<String> listIdsByHcParty(String healthcarePartyId);
	List<String> listIdsByHcPartyAndDateOfBirth(Integer date, String healthcarePartyId);
	List<String> listIdsByHcPartyAndDateOfBirth(Integer startDate, Integer endDate, String healthcarePartyId);
	List<String> listIdsForHcPartyDateOfBirth(Integer date, String healthcarePartyId);
	List<String> listIdsByHcPartyAndNameContainsFuzzy(String searchString, String healthcarePartyId);
	List<String> listIdsOfHcPartyNameContainsFuzzy(String searchString, String healthcarePartyId);

	List<String> listIdsByHcPartyAndExternalId(String externalId, String healthcarePartyId);

	PaginatedList<String> findIdsByHcParty(String healthcarePartyId, PaginationOffset pagination);
	PaginatedList<Patient> findPatientsByHcPartyAndName(String name, String healthcarePartyId, PaginationOffset pagination, boolean descending);
	PaginatedList<Patient> findPatientsOfHcPartyAndName(String name, String healthcarePartyId, PaginationOffset offset, boolean descending);
	PaginatedList<Patient> findPatientsByHcPartyAndSsin(String ssin, String healthcarePartyId, PaginationOffset pagination, boolean descending);
	PaginatedList<Patient> findPatientsOfHcPartyAndSsin(String ssin, String healthcarePartyId, PaginationOffset offset, boolean descending);

	PaginatedList<Patient> findPatientsByHcPartyModificationDate(Long startDate, Long endDate, String healthcarePartyId, PaginationOffset pagination, boolean descending);

	PaginatedList<Patient> findPatientsOfHcPartyModificationDate(Long startDate, Long endDate, String healthcarePartyId, PaginationOffset pagination, boolean descending);

	PaginatedList<Patient> findPatientsByHcPartyDateOfBirth(Integer startDate, Integer endDate, String healthcarePartyId, PaginationOffset pagination, boolean descending);
	PaginatedList<Patient> findPatientsOfHcPartyDateOfBirth(Integer startDate, Integer endDate, String healthcarePartyId, PaginationOffset pagination, boolean descending);


	Patient findPatientsByUserId(String id);
	List<Patient> get(Collection<String> patIds);

	Patient getByExternalId(String externalId);

    PaginatedList<Patient> findDeletedPatientsByDeleteDate(Long start, Long end, boolean descending, PaginationOffset paginationOffset);

	List<Patient> findDeletedPatientsByNames(String firstName, String lastName);

	List<Patient> listConflicts();

	PaginatedList<Patient> listOfPatientsModifiedAfter(Long date, PaginationOffset<Long> paginationOffset);

	List<String> listIdsByHcPartyAndSsins(Collection<String> ssins, String healthcarePartyId);

    List<String> listByHcPartyName(String searchString, String healthcarePartyId);
}

File: src/main/kotlin/org/taktik/icure/asyncdao/PatientDAO.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */

package org.taktik.icure.asyncdao

import kotlinx.coroutines.flow.Flow
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.couchdb.entity.ComplexKey
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.entities.Patient
import org.taktik.icure.entities.embed.Gender
import org.taktik.icure.entities.embed.Identifier

interface PatientDAO : GenericDAO<Patient> {

	fun listPatientIdsByHcPartyAndName(name: String, healthcarePartyId: String): Flow<String>
	fun listPatientIdsOfHcPartyAndName(name: String, healthcarePartyId: String): Flow<String>
	fun listPatientIdsByHcPartyAndSsin(ssin: String, healthcarePartyId: String): Flow<String>
	fun listPatientIdsOfHcPartyAndSsin(ssin: String, healthcarePartyId: String): Flow<String>
	fun listPatientIdsByActive(active: Boolean, healthcarePartyId: String): Flow<String>
	fun listOfMergesAfter(date: Long?): Flow<Patient>
	suspend fun countByHcParty(healthcarePartyId: String): Int
	suspend fun countOfHcParty(healthcarePartyId: String): Int
	fun listPatientIdsByHcParty(healthcarePartyId: String): Flow<String>
	fun listPatientIdsByHcPartyAndDateOfBirth(date: Int?, healthcarePartyId: String): Flow<String>

	fun listPatientIdsByHcPartyAndDateOfBirth(startDate: Int?, endDate: Int?, healthcarePartyId: String): Flow<String>
	fun listPatientIdsByHcPartyGenderEducationProfession(healthcarePartyId: String, gender: Gender?, education: String?, profession: String?): Flow<String>
	fun listPatientIdsForHcPartyDateOfBirth(date: Int?, healthcarePartyId: String): Flow<String>
	fun listPatientIdsByHcPartyAndNameContainsFuzzy(searchString: String?, healthcarePartyId: String, limit: Int?): Flow<String>
	fun listPatientIdsOfHcPartyNameContainsFuzzy(searchString: String?, healthcarePartyId: String, limit: Int?): Flow<String>

	fun listPatientIdsByHcPartyAndExternalId(externalId: String?, healthcarePartyId: String): Flow<String>

	fun findPatientIdsByHcParty(healthcarePartyId: String, pagination: PaginationOffset<ComplexKey>): Flow<ViewQueryResultEvent>
	fun findPatientsByHcPartyAndName(name: String?, healthcarePartyId: String, pagination: PaginationOffset<ComplexKey>, descending: Boolean): Flow<ViewQueryResultEvent>
	fun findPatientsOfHcPartyAndName(name: String?, healthcarePartyId: String, offset: PaginationOffset<ComplexKey>, descending: Boolean): Flow<ViewQueryResultEvent>
	fun findPatientsByHcPartyAndSsin(ssin: String?, healthcarePartyId: String, pagination: PaginationOffset<ComplexKey>, descending: Boolean): Flow<ViewQueryResultEvent>
	fun findPatientsOfHcPartyAndSsin(ssin: String?, healthcarePartyId: String, offset: PaginationOffset<ComplexKey>, descending: Boolean): Flow<ViewQueryResultEvent>

	fun findPatientsByHcPartyModificationDate(startDate: Long?, endDate: Long?, healthcarePartyId: String, pagination: PaginationOffset<ComplexKey>, descending: Boolean): Flow<ViewQueryResultEvent>

	fun findPatientsOfHcPartyModificationDate(startDate: Long?, endDate: Long?, healthcarePartyId: String, pagination: PaginationOffset<ComplexKey>, descending: Boolean): Flow<ViewQueryResultEvent>

	fun findPatientsByHcPartyDateOfBirth(startDate: Int?, endDate: Int?, healthcarePartyId: String, pagination: PaginationOffset<ComplexKey>, descending: Boolean): Flow<ViewQueryResultEvent>
	fun findPatientsOfHcPartyDateOfBirth(startDate: Int?, endDate: Int?, healthcarePartyId: String, pagination: PaginationOffset<ComplexKey>, descending: Boolean): Flow<ViewQueryResultEvent>

	suspend fun findPatientsByUserId(id: String): Patient?
	fun getPatients(patIds: Collection<String>): Flow<Patient>

	suspend fun getPatientByExternalId(externalId: String): Patient?

	fun findDeletedPatientsByDeleteDate(start: Long, end: Long?, descending: Boolean, paginationOffset: PaginationOffset<Long>): Flow<ViewQueryResultEvent>

	fun findDeletedPatientsByNames(firstName: String?, lastName: String?): Flow<Patient>

	fun listConflicts(): Flow<Patient>

	fun findPatientsModifiedAfter(date: Long, paginationOffset: PaginationOffset<Long>): Flow<ViewQueryResultEvent>

	fun listPatientIdsByHcPartyAndSsins(ssins: Collection<String>, healthcarePartyId: String): Flow<String>

	fun listPatientsByHcPartyName(searchString: String?, healthcarePartyId: String): Flow<String>

	suspend fun getHcPartyKeysForDelegate(healthcarePartyId: String): Map<String, String>

	fun getDuplicatePatientsBySsin(healthcarePartyId: String, paginationOffset: PaginationOffset<ComplexKey>): Flow<ViewQueryResultEvent>

	fun getDuplicatePatientsByName(healthcarePartyId: String, paginationOffset: PaginationOffset<ComplexKey>): Flow<ViewQueryResultEvent>

	fun findPatients(ids: Collection<String>): Flow<ViewQueryResultEvent>

	fun findPatients(ids: Flow<String>): Flow<ViewQueryResultEvent>

	fun listPatientIdsByHcPartyAndIdentifiers(healthcarePartyId: String, identifiers: List<Identifier>): Flow<String>

	fun listPatientsByHcPartyAndIdentifier(healthcarePartyId: String, system: String, id: String): Flow<Patient>
}
